docs/advanced-tool-implementation.md

```diff
@@ -0,0 +1,164 @@
+
+!!! note
+    The advanced tool API is multiplatform. This lets you use the same tools across different platforms.
+
+## Tool implementation
+
+The Koog framework provides the following approaches for implementing tools:
+
+* Using the base class `Tool` for all tools. You should use this class when you need to return non-text results or require complete control over the tool behavior.
+* Using the `SimpleTool` class that extends the base `Tool` class and simplifies the creation of tools that return text results. You should use this approach for scenarios where the 
+  tool only needs to return a text.
+
+Both approaches use the same core components but differ in implementation and the results they return.
+
+### Tool class
+
+The [`Tool<Args, Result>`](https://api.koog.ai/agents/agents-tools/ai.koog.agents.core.tools/-tool/index.html) abstract class is the base class for creating tools in Koog.
+It lets you create tools that accept specific argument types (`Args`) and return results of various types (`Result`).
+
+Each tool consists of the following components:
+
+| <div style="width:110px">Component</div> | Description                                                                                                                                                                                                                                                                                                                   |
+|------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `Args`                                   | The serializable data class that defines arguments required for the tool. This class must implement the [`Tool.Args`](https://api.koog.ai/agents/agents-tools/ai.koog.agents.core.tools/-tool/-args/index.html) interface. For tools that do not require arguments, you can use the built-in `ToolArgs.Empty` implementation. |
+| `Result`                                 | The type of result that the tool returns. This must implement the [`ToolResult`](https://api.koog.ai/agents/agents-tools/ai.koog.agents.core.tools/-tool-result/index.html) interface, which can be `ToolResult.Text`, `ToolResult.Boolean`, `ToolResult.Number`, or a custom implementation of `ToolResult.JSONSerializable`. |
+| `argsSerializer`                         | The overridden variable that defines how the arguments for the tool are serialized. See also [argsSerializer](https://api.koog.ai/agents/agents-tools/ai.koog.agents.core.tools/-tool/args-serializer.html).                                                                                                                  |
```

Comment:
It's better to write "deserialized" instead of "serialized" here, since we use this primarily for serialization from string to class object